### DISCUSSION TITLE
Combining OpenAPI overriding with protected Docs

### DISCUSSION THREAD

**Discussion by usl-cto** [fastapi/fastapi#9323] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import secrets

from fastapi import FastAPI, Depends, HTTPException
from fastapi.openapi.docs import get_swagger_ui_html, get_redoc_html
from fastapi.openapi.utils import get_openapi
from fastapi.security import HTTPBasicCredentials, HTTPBasic
from starlette import status

from api.entrypoints.fastapi.routers.files import m_files, p_files
from api.entrypoints.fastapi.routers.internal import ping

app = FastAPI(
    title="App",
    version="2.24.0",
    redoc_url=None,
    docs_url=None,
    openapi_url=None,
    swagger_ui_oauth2_redirect_url=None,
    swagger_ui_parameters={"operationsSorter": "method"},
)

security = HTTPBasic()


def get_current_username(credentials: HTTPBasicCredentials = Depends(security)):
    # TODO this is just a placeholder for now
    correct_username = secrets.compare_digest(credentials.username, "user")
    correct_password = secrets.compare_digest(credentials.password, "password")
    if not (correct_username and correct_password):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Incorrect email or password",
            headers={"WWW-Authenticate": "Basic"},
        )
    return credentials.username


@app.get("/docs")
async def get_swagger_documentation(username: str = Depends(get_current_username)):
    return get_swagger_ui_html(openapi_url="/openapi.json", title="docs")


@app.get("/redoc")
async def get_redoc_documentation(username: str = Depends(get_current_username)):
    return get_redoc_html(openapi_url="/openapi.json", title="docs")


@app.get("/openapi.json", include_in_schema=False)
async def openapi(username: str = Depends(get_current_username)):
    return get_openapi(title=app.title, version=app.version, routes=app.routes)


app.include_router(ping.router)
app.include_router(m_files.router)
app.include_router(p_files.router)


def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema

    openapi_schema = get_openapi(
        title="App",
        version="2.25.0",
        description="This is a very custom OpenAPI schema",
        routes=app.routes,
    )

    # set the x-logo object for all paths
    openapi_schema["info"]["x-logo"] = {"url": "https://fastapi.tiangolo.com/img/logo-margin/logo-teal.png"}

    app.openapi_schema = openapi_schema

    return app.openapi_schema


app.openapi = custom_openapi
```

### Description

Basically I want to Override the docs as described here https://fastapi.tiangolo.com/advanced/extending-openapi/#overriding-the-defaults to add a custom logo. That, I can get it working.

Additionally, I also want to add auth to the docs (basic is ok for now) like it's explained here:
originally posted by @kennylajara in https://github.com/tiangolo/fastapi/discussions/8169#discussioncomment-5147902
And I also got it working on its own.

The problem is, that I'm not able to make them work together at the same time. 

### Operating System

Linux

### FastAPI Version

0.95.0

### Python Version

3.8.16

**Comment by usl-cto** [comment]
I've progressed a little bit on this:

```python
from fastapi import FastAPI, Depends
from fastapi.openapi.utils import get_openapi

from api.entrypoints.fastapi.constants import TAGS_METADATA, API_DESCRIPTION, OPENAPI_PATH, FAVICON_URL
from api.entrypoints.fastapi.routers.docs import docs
from api.entrypoints.fastapi.routers.files import m_files, p_files
from api.entrypoints.fastapi.routers.internal import ping
from api.entrypoints.fastapi.util import get_current_username

app = FastAPI(
    swagger_ui_parameters={"operationsSorter": "method", "tryItOutEnabled": "true"},
    openapi_tags=TAGS_METADATA,
    redoc_url=None,
    docs_url=None,
    # openapi_url=None, # TODO if I include this, to later protect it, the rest of attributes won't work
)


@app.get(OPENAPI_PATH, include_in_schema=False)
async def openapi(username: str = Depends(get_current_username)):
    return get_openapi(title=app.title, version=app.version, routes=app.routes)


@app.get("/favicon.ico", include_in_schema=False)
async def favicon():
    return FAVICON_URL


app.include_router(docs.router)
app.include_router(ping.router)
app.include_router(m_files.router)
app.include_router(p_files.router)


def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema

    openapi_schema = get_openapi(
        title="App",
        description=API_DESCRIPTION,
        version="2.24.0",
        terms_of_service="URL_TO_TERMS",
        contact={
            "name": "Development team @ company",
            "url": "URL_TO_CONTACT",
            "email": "dev@company.com",
        },
        license_info={
            "name": "Copyright © dates Company",
            "url": "company.com/legal-warning/",
        },
        servers=[
            {"url": "http://localhost:8000", "description": "Development Server"},
            {
                "url": "https://mock.pstmn.io",
                "description": "Mock Server",
            },
        ],
        routes=app.routes,
    )

    # set the x-logo object for all paths
    openapi_schema["info"]["x-logo"] = {"url": "LOGO_URL"}

    app.openapi_schema = openapi_schema

    return app.openapi_schema


app.openapi = custom_openapi
```

With this code I'm able to setup all the attributes in the OpenAPI file. But, if I try to protect the openapi.json path so that nobody can access it (except for FastAPI itself, that's what I intend to do), then none of the attributes that I set up will be visible.

I understand that as  I'm protecting the OpenAPI path, FastAPI is not able to access it and therefore cannot load it for the docs.

Any idea on how to protect `openapi.json` from being accessed externaly?

**Reply by usl-cto** [reply]
May be related: https://github.com/tiangolo/fastapi/discussions/6449
